File: bonfire_lib/k8s_client.py
"""Kubernetes API client for ephemeral resource CRDs.

Uses the kubernetes Python client directly (no ocviapy/oc dependency).
Supports three auth modes: explicit server+token, in-cluster, kubeconfig.
"""

import atexit
import base64
import copy
import json
import logging
import os
import tempfile

from kubernetes import client, config
from kubernetes.client import ApisApi, ApiException
from kubernetes.config import ConfigException
from kubernetes.dynamic import DynamicClient

log = logging.getLogger(__name__)

CRD_API_VERSION = "cloud.redhat.com/v1alpha1"

DEFAULT_READ_TIMEOUT = 30
DEFAULT_WRITE_TIMEOUT = 60


def _sanitize_username(name: str) -> str:
    """Sanitize a username for use as a K8s label value.

    Matches bonfire's existing sanitization in openshift.py:whoami().
    """
    return name.replace("@", "_at_").replace(":", "_")


def _extract_username(context_user: str) -> str:
    """Extract the username from a kubeconfig context user string.

    Kubeconfig context user strings often include the cluster API URL
    (e.g., 'gbuchana/api-crc-eph-r9lp-p1-openshiftapps-com:6443').
    This extracts just the username part before the first '/'.
    """
    if "/" in context_user:
        return context_user.split("/")[0]
    return context_user


class EphemeralK8sClient:
    """Kubernetes API client for ephemeral resource CRDs.

    Supports three auth modes (auto-detected in priority order):
    1. Explicit server + token
    2. In-cluster service account
    3. Kubeconfig file
    """

    def __init__(
        self,
        kubeconfig_path: str | None = None,
        context: str | None = None,
        server: str | None = None,
        token: str | None = None,
        ca_data: str | None = None,
        skip_tls: bool = False,
    ):
        self._auth_mode = "kubeconfig"
        if server and token:
            self._auth_mode = "token"
            configuration = client.Configuration()
            configuration.host = server
            configuration.api_key = {"authorization": f"Bearer {token}"}
            if skip_tls:
                configuration.verify_ssl = False
            elif ca_data:
                ca_file = tempfile.NamedTemporaryFile(delete=False, suffix=".crt")
                ca_file.write(base64.b64decode(ca_data))
                ca_file.close()
                configuration.ssl_ca_cert = ca_file.name
                atexit.register(os.unlink, ca_file.name)
            self._api_client = client.ApiClient(configuration)
        else:
            # Try kubeconfig first, then fall back to in-cluster auth.
            # This ensures that explicit `oc login` credentials are used even when
            # running inside a pod (e.g., GitLab CI runners).
            try:
                # Use explicit configuration instead of relying on global state
                kube_config = client.Configuration()
                config.load_kube_config(
                    config_file=kubeconfig_path,
                    context=context,
                    client_configuration=kube_config,
                )
                # Verify that the kubeconfig actually works by probing /apis
                test_client = client.ApiClient(kube_config)
                ApisApi(test_client).get_api_versions()
                # If we get here, kubeconfig is valid and works
                self._auth_mode = "kubeconfig"
                self._api_client = test_client
            except (ConfigException, ApiException) as e:
                # Kubeconfig doesn't exist, is invalid, or doesn't have working credentials
                log.error(
                    "Kubeconfig auth failed (%s: %s), attempting in-cluster fallback",
                    e.__class__.__name__,
                    str(e)[:100],
                )

                if self._is_in_cluster():
                    self._auth_mode = "in-cluster"
                    incluster_config = client.Configuration()
                    config.load_incluster_config(client_configuration=incluster_config)
                    self._api_client = client.ApiClient(incluster_config)
                else:
                    raise ConfigException(
                        "Unable to load kubeconfig and not running in-cluster"
                    ) from e

        self._dynamic = DynamicClient(self._api_client)
        self._core_v1 = client.CoreV1Api(self._api_client)

    @staticmethod
    def _is_in_cluster() -> bool:
        """Check if running inside a Kubernetes pod."""
        return os.path.exists("/var/run/secrets/kubernetes.io/serviceaccount/token")

    def _get_resource(self, kind: str):
        """Get a DynamicClient resource handle for a cloud.redhat.com/v1alpha1 CRD."""
        try:
            return self._dynamic.resources.get(api_version=CRD_API_VERSION, kind=kind)
        except Exception as e:
            # Log available resources for debugging
            log.error(
                f"Failed to get resource {kind} from {CRD_API_VERSION}. "
                f"Error: {e.__class__.__name__}: {e}"
            )

    # --- NamespaceReservation operations ---

    def create_reservation(self, body: dict) -> dict:
        """Create a NamespaceReservation CR."""
        resource = self._get_resource("NamespaceReservation")
        return resource.create(body=body, _request_timeout=DEFAULT_WRITE_TIMEOUT).to_dict()

    def get_reservation(self, name: str) -> dict | None:
        """Get a NamespaceReservation by name. Returns None if not found."""
        resource = self._get_resource("NamespaceReservation")
        try:
            return resource.get(name=name, _request_timeout=DEFAULT_READ_TIMEOUT).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    def list_reservations(self, label_selector: str | None = None) -> list[dict]:
        """List all NamespaceReservation CRs."""
        resource = self._get_resource("NamespaceReservation")
        kwargs = {"_request_timeout": DEFAULT_READ_TIMEOUT}
        if label_selector:
            kwargs["label_selector"] = label_selector
        return [item.to_dict() for item in resource.get(**kwargs).items]

    def patch_reservation(self, name: str, body: dict) -> dict:
        """Patch a NamespaceReservation CR (merge patch)."""
        resource = self._get_resource("NamespaceReservation")
        return resource.patch(
            name=name,
            body=body,
            content_type="application/merge-patch+json",
            _request_timeout=DEFAULT_WRITE_TIMEOUT,
        ).to_dict()

    # --- NamespacePool operations ---

    def list_pools(self) -> list[dict]:
        """List all NamespacePool CRs."""
        resource = self._get_resource("NamespacePool")
        return [
            item.to_dict() for item in resource.get(_request_timeout=DEFAULT_READ_TIMEOUT).items
        ]

    def get_pool(self, name: str) -> dict | None:
        """Get a NamespacePool by name."""
        resource = self._get_resource("NamespacePool")
        try:
            return resource.get(name=name, _request_timeout=DEFAULT_READ_TIMEOUT).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    # --- ClusterReservation operations ---

    def get_cluster_reservation(self, name: str) -> dict | None:
        """Get a ClusterReservation by name. Returns None if not found."""
        resource = self._get_resource("ClusterReservation")
        try:
            return resource.get(name=name, _request_timeout=DEFAULT_READ_TIMEOUT).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    # --- Namespace operations (via CoreV1Api) ---

    def get_namespace(self, name: str) -> dict | None:
        """Get a namespace by name."""
        try:
            return self._core_v1.read_namespace(
                name=name, _request_timeout=DEFAULT_READ_TIMEOUT
            ).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    def get_configmap(self, name: str, namespace: str) -> dict | None:
        """Get a ConfigMap by name and namespace."""
        try:
            return self._core_v1.read_namespaced_config_map(
                name=name, namespace=namespace, _request_timeout=DEFAULT_READ_TIMEOUT
            ).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    def get_secret(self, name: str, namespace: str) -> dict | None:
        """Get a Secret by name and namespace."""
        try:
            return self._core_v1.read_namespaced_secret(
                name=name, namespace=namespace, _request_timeout=DEFAULT_READ_TIMEOUT
            ).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    def list_namespaces(self, label_selector: str | None = None) -> list[dict]:
        """List namespaces, optionally filtered by label."""
        kwargs = {"_request_timeout": DEFAULT_READ_TIMEOUT}
        if label_selector:
            kwargs["label_selector"] = label_selector
        return [item.to_dict() for item in self._core_v1.list_namespace(**kwargs).items]

    # --- Generic CRD operations (for ClowdApp, Frontend, etc.) ---

    def list_crds(self, kind: str, namespace: str | None = None) -> list[dict]:
        """List CRDs of a given kind, optionally in a namespace."""
        resource = self._get_resource(kind)
        kwargs = {"_request_timeout": DEFAULT_READ_TIMEOUT}
        if namespace:
            kwargs["namespace"] = namespace
        return [item.to_dict() for item in resource.get(**kwargs).items]

    def get_crd(self, kind: str, name: str, namespace: str | None = None) -> dict | None:
        """Get a single CRD by kind and name."""
        resource = self._get_resource(kind)
        try:
            kwargs = {"name": name, "_request_timeout": DEFAULT_READ_TIMEOUT}
            if namespace:
                kwargs["namespace"] = namespace
            return resource.get(**kwargs).to_dict()
        except ApiException as e:
            if e.status == 404:
                return None
            raise

    # --- Generic resource operations (for deploy flow) ---

    def get_dynamic_resource(self, api_version: str, kind: str):
        """Get a DynamicClient resource handle for any API version/kind."""
        return self._dynamic.resources.get(api_version=api_version, kind=kind)

    def apply_resource(self, body: dict, namespace: str | None = None) -> dict:
        """Create or update a K8s resource.

        Tries GET first; creates on 404, patches on existing.
        """
        api_version = body.get("apiVersion", "")
        kind = body.get("kind", "")
        name = body.get("metadata", {}).get("name", "")
        ns = namespace or body.get("metadata", {}).get("namespace")

        resource = self.get_dynamic_resource(api_version, kind)

        get_kwargs = {"name": name, "_request_timeout": DEFAULT_READ_TIMEOUT}
        if ns:
            get_kwargs["namespace"] = ns

        try:
            resource.get(**get_kwargs)
            patch_kwargs = {
                "name": name,
                "body": body,
                "content_type": "application/merge-patch+json",
                "_request_timeout": DEFAULT_WRITE_TIMEOUT,
            }
            if ns:
                patch_kwargs["namespace"] = ns
            return resource.patch(**patch_kwargs).to_dict()
        except ApiException as e:
            if e.status == 404:
                create_kwargs = {
                    "body": body,
                    "_request_timeout": DEFAULT_WRITE_TIMEOUT,
                }
                if ns:
                    create_kwargs["namespace"] = ns
                return resource.create(**create_kwargs).to_dict()
            raise

    def list_dynamic_resources(
        self, api_version: str, kind: str, namespace: str | None = None
    ) -> list[dict]:
        """List resources of any API version/kind."""
        resource = self.get_dynamic_resource(api_version, kind)
        kwargs = {"_request_timeout": DEFAULT_READ_TIMEOUT}
        if namespace:
            kwargs["namespace"] = namespace
        return [item.to_dict() for item in resource.get(**kwargs).items]

    def process_template(
        self, template: dict, parameters: dict, namespace: str = "default"
    ) -> list[dict]:
        """Process an OpenShift Template via the processedtemplates API.

        Fills parameter values into the template, POSTs to the cluster's
        template.openshift.io/v1 processedtemplates endpoint, and returns
        the expanded objects list.

        Args:
            template: OpenShift Template dict (kind: Template)
            parameters: Parameter name->value pairs to substitute
            namespace: Namespace for the processedtemplates API call

        Returns:
            List of processed K8s resource dicts
        """
        template = copy.deepcopy(template)

        if template.get("apiVersion") == "v1" and template.get("kind") == "Template":
            template["apiVersion"] = "template.openshift.io/v1"

        for param_def in template.get("parameters", []):
            name = param_def["name"]
            if name in parameters:
                val = parameters[name]
                if isinstance(val, bool):
                    val = str(val).lower()
                param_def["value"] = str(val)

        path = f"/apis/template.openshift.io/v1/namespaces/{namespace}/processedtemplates"

        response = self._api_client.call_api(
            path,
            "POST",
            body=template,
            response_types_map={200: "object", 201: "object"},
            _request_timeout=DEFAULT_WRITE_TIMEOUT,
            header_params={
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
            auth_settings=["BearerToken"],
        )

        processed = response[0]
        return processed.get("objects", [])

    # --- Identity ---

    def whoami(self) -> str:
        """Get the current authenticated user identity.

        For token auth: uses TokenReview to resolve the token's identity.
        For kubeconfig auth: reads the user from the active context.
        For in-cluster auth: uses TokenReview with the service account token.
        Falls back to 'unknown' if identity cannot be determined.

        The returned name is sanitized for use as a K8s label value:
        '@' -> '_at_', ':' -> '_'
        """
        if self._auth_mode == "kubeconfig":
            try:
                _, active_context = config.list_kube_config_contexts()
                if active_context:
                    user = active_context.get("context", {}).get("user", "")
                    if user:
                        return _sanitize_username(_extract_username(user))
            except config.ConfigException:
                pass

        try:
            auth_v1 = client.AuthenticationV1Api(self._api_client)
            review = auth_v1.create_token_review(
                body=client.V1TokenReview(
                    spec=client.V1TokenReviewSpec(
                        token=self._api_client.configuration.api_key.get(
                            "authorization", ""
                        ).replace("Bearer ", "")
                    )
                ),
                _request_timeout=DEFAULT_READ_TIMEOUT,
            )
            username = review.status.user.username
            if username:
                return _sanitize_username(username)
        except Exception:
            pass

        return "unknown"

File: go/pkg/ephemeral/k8s_client.go
package ephemeral

import (
	"context"
	"encoding/base64"
	"fmt"
	"log/slog"
	"os"

	authv1 "k8s.io/api/authentication/v1"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)

const (
	crdGroup   = "cloud.redhat.com"
	crdVersion = "v1alpha1"

	kindNamespaceReservation = "NamespaceReservation"
	kindNamespacePool        = "NamespacePool"
	kindClusterReservation   = "ClusterReservation"
	kindClusterPool          = "ClusterPool"
	kindClowdApp             = "ClowdApp"
	kindFrontend             = "Frontend"
	kindFrontendEnvironment  = "FrontendEnvironment"

	defaultReadTimeout  = 30
	defaultWriteTimeout = 60
)

var (
	gvrNamespaceReservation = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "namespacereservations"}
	gvrNamespacePool        = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "namespacepools"}
	gvrClusterReservation   = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "clusterreservations"}
	gvrClusterPool          = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "clusterpools"}
	gvrClowdApp             = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "clowdapps"}
	gvrFrontend             = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "frontends"}
	gvrFrontendEnvironment  = schema.GroupVersionResource{Group: crdGroup, Version: crdVersion, Resource: "frontendenvironments"}
)

// AuthMode describes how the client authenticated to the cluster.
type AuthMode string

const (
	AuthModeToken     AuthMode = "token"
	AuthModeInCluster AuthMode = "in-cluster"
	AuthModeKubeconfig AuthMode = "kubeconfig"
)

// ClientOptions holds optional parameters for NewClient.
type ClientOptions struct {
	// KubeconfigPath overrides the default kubeconfig file location.
	KubeconfigPath string
	// Context selects a specific kubeconfig context.
	Context string
	// Server is the API server URL for explicit token auth.
	Server string
	// Token is the bearer token for explicit token auth.
	Token string
	// CAData is base64-encoded CA certificate data for token auth.
	CAData string
	// SkipTLS disables TLS verification (token auth only).
	SkipTLS bool
}

// EphemeralK8sClient is the Kubernetes API client for ephemeral resource CRDs.
// It mirrors bonfire_lib/k8s_client.py:EphemeralK8sClient.
//
// Auth modes (auto-detected in priority order):
//  1. Explicit Server + Token (ClientOptions.Server + ClientOptions.Token)
//  2. In-cluster service account (/var/run/secrets/kubernetes.io/serviceaccount/token)
//  3. Kubeconfig file (KUBECONFIG env var or ~/.kube/config)
type EphemeralK8sClient struct {
	authMode   AuthMode
	dynamic    dynamic.Interface
	core       kubernetes.Interface
	restConfig *rest.Config
}

// NewClient creates an EphemeralK8sClient using the provided options.
// Auth mode is selected in the priority order documented on EphemeralK8sClient.
func NewClient(opts ClientOptions) (*EphemeralK8sClient, error) {
	var cfg *rest.Config
	var authMode AuthMode
	var err error

	if opts.Server != "" && opts.Token != "" {
		// Priority 1: explicit server + token
		authMode = AuthModeToken
		cfg = &rest.Config{
			Host:        opts.Server,
			BearerToken: opts.Token,
		}
		if opts.SkipTLS {
			cfg.TLSClientConfig = rest.TLSClientConfig{Insecure: true}
		} else if opts.CAData != "" {
			caBytes, err2 := base64.StdEncoding.DecodeString(opts.CAData)
			if err2 != nil {
				return nil, fmt.Errorf("invalid CA data: %w", err2)
			}
			cfg.TLSClientConfig = rest.TLSClientConfig{CAData: caBytes}
		}
	} else {
		// Try kubeconfig first (Priority 3), fall back to in-cluster (Priority 2).
		// This mirrors the Python behaviour: prefer explicit oc-login credentials even
		// when running inside a pod (e.g. GitLab CI runners).
		loadRules := clientcmd.NewDefaultClientConfigLoadingRules()
		if opts.KubeconfigPath != "" {
			loadRules.ExplicitPath = opts.KubeconfigPath
		}
		overrides := &clientcmd.ConfigOverrides{}
		if opts.Context != "" {
			overrides.CurrentContext = opts.Context
		}
		kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadRules, overrides)
		cfg, err = kubeConfig.ClientConfig()
		if err != nil {
			// Kubeconfig failed — try in-cluster
			if isInCluster() {
				slog.Warn("kubeconfig auth failed, attempting in-cluster fallback", "err", err)
				cfg, err = rest.InClusterConfig()
				if err != nil {
					return nil, fmt.Errorf("both kubeconfig and in-cluster auth failed: %w", err)
				}
				authMode = AuthModeInCluster
			} else {
				return nil, fmt.Errorf("unable to load kubeconfig and not running in-cluster: %w", err)
			}
		} else {
			authMode = AuthModeKubeconfig
			// Verify the kubeconfig actually works by hitting /api
			testClient, err2 := kubernetes.NewForConfig(cfg)
			if err2 != nil {
				return nil, fmt.Errorf("failed to build k8s client: %w", err2)
			}
			if _, err2 = testClient.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{Limit: 1}); err2 != nil {
				if isInCluster() {
					slog.Warn("kubeconfig connectivity check failed, attempting in-cluster fallback", "err", err2)
					cfg, err = rest.InClusterConfig()
					if err != nil {
						return nil, fmt.Errorf("both kubeconfig and in-cluster auth failed: %w", err)
					}
					authMode = AuthModeInCluster
				} else {
					return nil, fmt.Errorf("kubeconfig auth failed: %w", err2)
				}
			}
		}
	}

	dynClient, err := dynamic.NewForConfig(cfg)
	if err != nil {
		return nil, fmt.Errorf("failed to create dynamic client: %w", err)
	}
	coreClient, err := kubernetes.NewForConfig(cfg)
	if err != nil {
		return nil, fmt.Errorf("failed to create core client: %w", err)
	}

	return &EphemeralK8sClient{
		authMode:   authMode,
		dynamic:    dynClient,
		core:       coreClient,
		restConfig: cfg,
	}, nil
}

// isInCluster returns true if running inside a Kubernetes pod.
func isInCluster() bool {
	_, err := os.Stat("/var/run/secrets/kubernetes.io/serviceaccount/token")
	return err == nil
}

// AuthMode returns the auth mode used by this client.
func (c *EphemeralK8sClient) AuthMode() AuthMode { return c.authMode }

// --- helpers ---

func ignoreNotFound(err error) error {
	if errors.IsNotFound(err) {
		return nil
	}
	return err
}

func (c *EphemeralK8sClient) gvrResource(gvr schema.GroupVersionResource) dynamic.ResourceInterface {
	return c.dynamic.Resource(gvr)
}

func (c *EphemeralK8sClient) gvrNamespacedResource(gvr schema.GroupVersionResource, namespace string) dynamic.ResourceInterface {
	return c.dynamic.Resource(gvr).Namespace(namespace)
}

// --- NamespaceReservation operations ---

// CreateReservation creates a NamespaceReservation CR.
func (c *EphemeralK8sClient) CreateReservation(ctx context.Context, body map[string]any) (map[string]any, error) {
	obj := &unstructured.Unstructured{Object: body}
	result, err := c.gvrResource(gvrNamespaceReservation).Create(ctx, obj, metav1.CreateOptions{})
	if err != nil {
		return nil, fmt.Errorf("create NamespaceReservation: %w", err)
	}
	return result.Object, nil
}

// GetReservation returns a NamespaceReservation by name. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetReservation(ctx context.Context, name string) (map[string]any, error) {
	result, err := c.gvrResource(gvrNamespaceReservation).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return nil, ignoreNotFound(err)
	}
	return result.Object, nil
}

// ListReservations lists all NamespaceReservation CRs, optionally filtered by labelSelector.
func (c *EphemeralK8sClient) ListReservations(ctx context.Context, labelSelector string) ([]map[string]any, error) {
	opts := metav1.ListOptions{LabelSelector: labelSelector}
	list, err := c.gvrResource(gvrNamespaceReservation).List(ctx, opts)
	if err != nil {
		return nil, fmt.Errorf("list NamespaceReservations: %w", err)
	}
	return unstructuredItems(list), nil
}

// PatchReservation applies a merge patch to a NamespaceReservation.
func (c *EphemeralK8sClient) PatchReservation(ctx context.Context, name string, patch map[string]any) (map[string]any, error) {
	data, err := marshalJSON(patch)
	if err != nil {
		return nil, err
	}
	result, err := c.gvrResource(gvrNamespaceReservation).Patch(ctx, name, types.MergePatchType, data, metav1.PatchOptions{})
	if err != nil {
		return nil, fmt.Errorf("patch NamespaceReservation %q: %w", name, err)
	}
	return result.Object, nil
}

// --- NamespacePool operations ---

// ListPools lists all NamespacePool CRs.
func (c *EphemeralK8sClient) ListPools(ctx context.Context) ([]map[string]any, error) {
	list, err := c.gvrResource(gvrNamespacePool).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list NamespacePools: %w", err)
	}
	return unstructuredItems(list), nil
}

// GetPool returns a NamespacePool by name. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetPool(ctx context.Context, name string) (map[string]any, error) {
	result, err := c.gvrResource(gvrNamespacePool).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return nil, ignoreNotFound(err)
	}
	return result.Object, nil
}

// --- ClusterReservation operations ---

// CreateClusterReservation creates a ClusterReservation CR.
func (c *EphemeralK8sClient) CreateClusterReservation(ctx context.Context, body map[string]any) (map[string]any, error) {
	obj := &unstructured.Unstructured{Object: body}
	result, err := c.gvrResource(gvrClusterReservation).Create(ctx, obj, metav1.CreateOptions{})
	if err != nil {
		return nil, fmt.Errorf("create ClusterReservation: %w", err)
	}
	return result.Object, nil
}

// GetClusterReservation returns a ClusterReservation by name. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetClusterReservation(ctx context.Context, name string) (map[string]any, error) {
	result, err := c.gvrResource(gvrClusterReservation).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return nil, ignoreNotFound(err)
	}
	return result.Object, nil
}

// ListClusterReservations lists all ClusterReservation CRs, optionally filtered by labelSelector.
func (c *EphemeralK8sClient) ListClusterReservations(ctx context.Context, labelSelector string) ([]map[string]any, error) {
	opts := metav1.ListOptions{LabelSelector: labelSelector}
	list, err := c.gvrResource(gvrClusterReservation).List(ctx, opts)
	if err != nil {
		return nil, fmt.Errorf("list ClusterReservations: %w", err)
	}
	return unstructuredItems(list), nil
}

// PatchClusterReservation applies a merge patch to a ClusterReservation.
func (c *EphemeralK8sClient) PatchClusterReservation(ctx context.Context, name string, patch map[string]any) (map[string]any, error) {
	data, err := marshalJSON(patch)
	if err != nil {
		return nil, err
	}
	result, err := c.gvrResource(gvrClusterReservation).Patch(ctx, name, types.MergePatchType, data, metav1.PatchOptions{})
	if err != nil {
		return nil, fmt.Errorf("patch ClusterReservation %q: %w", name, err)
	}
	return result.Object, nil
}

// --- ClusterPool operations ---

// ListClusterPools lists all ClusterPool CRs.
func (c *EphemeralK8sClient) ListClusterPools(ctx context.Context) ([]map[string]any, error) {
	list, err := c.gvrResource(gvrClusterPool).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list ClusterPools: %w", err)
	}
	return unstructuredItems(list), nil
}

// GetClusterPool returns a ClusterPool by name. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetClusterPool(ctx context.Context, name string) (map[string]any, error) {
	result, err := c.gvrResource(gvrClusterPool).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return nil, ignoreNotFound(err)
	}
	return result.Object, nil
}

// --- Namespace operations (CoreV1Api) ---

// GetNamespace returns a namespace by name. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetNamespace(ctx context.Context, name string) (*corev1.Namespace, error) {
	ns, err := c.core.CoreV1().Namespaces().Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		if errors.IsNotFound(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("get namespace %q: %w", name, err)
	}
	return ns, nil
}

// GetConfigMap returns a ConfigMap by name and namespace. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetConfigMap(ctx context.Context, name, namespace string) (*corev1.ConfigMap, error) {
	cm, err := c.core.CoreV1().ConfigMaps(namespace).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		if errors.IsNotFound(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("get configmap %q/%q: %w", namespace, name, err)
	}
	return cm, nil
}

// GetSecret returns a Secret by name and namespace. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetSecret(ctx context.Context, name, namespace string) (*corev1.Secret, error) {
	secret, err := c.core.CoreV1().Secrets(namespace).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		if errors.IsNotFound(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("get secret %q/%q: %w", namespace, name, err)
	}
	return secret, nil
}

// --- Generic CRD operations ---

// ListCRDs lists CRDs of a given GVR, optionally scoped to a namespace.
func (c *EphemeralK8sClient) ListCRDs(ctx context.Context, gvr schema.GroupVersionResource, namespace string) ([]map[string]any, error) {
	var ri dynamic.ResourceInterface
	if namespace != "" {
		ri = c.dynamic.Resource(gvr).Namespace(namespace)
	} else {
		ri = c.dynamic.Resource(gvr)
	}
	list, err := ri.List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list %s: %w", gvr.Resource, err)
	}
	return unstructuredItems(list), nil
}

// GetCRD returns a single CRD by GVR and name, optionally in a namespace. Returns nil, nil if not found.
func (c *EphemeralK8sClient) GetCRD(ctx context.Context, gvr schema.GroupVersionResource, name, namespace string) (map[string]any, error) {
	var ri dynamic.ResourceInterface
	if namespace != "" {
		ri = c.dynamic.Resource(gvr).Namespace(namespace)
	} else {
		ri = c.dynamic.Resource(gvr)
	}
	result, err := ri.Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return nil, ignoreNotFound(err)
	}
	return result.Object, nil
}

// --- Identity ---

// Whoami returns the current authenticated user identity, sanitized for use
// as a K8s label value (@ → _at_, : → _).
// Falls back to "unknown" if identity cannot be determined.
func (c *EphemeralK8sClient) Whoami(ctx context.Context) string {
	if c.authMode == AuthModeKubeconfig {
		loadRules := clientcmd.NewDefaultClientConfigLoadingRules()
		kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadRules, &clientcmd.ConfigOverrides{})
		rawConfig, err := kubeConfig.RawConfig()
		if err == nil {
			if contextName := rawConfig.CurrentContext; contextName != "" {
				if ctx2, ok := rawConfig.Contexts[contextName]; ok && ctx2 != nil {
					user := ctx2.AuthInfo
					if user != "" {
						return sanitizeUsername(extractUsername(user))
					}
				}
			}
		}
	}

	// Token review for token/in-cluster modes
	token := c.restConfig.BearerToken
	if token == "" {
		return "unknown"
	}
	review, err := c.core.AuthenticationV1().TokenReviews().Create(ctx, &authv1.TokenReview{
		Spec: authv1.TokenReviewSpec{Token: token},
	}, metav1.CreateOptions{})
	if err != nil {
		slog.Debug("whoami token review failed", "err", err)
		return "unknown"
	}
	if review.Status.User.Username != "" {
		return sanitizeUsername(review.Status.User.Username)
	}
	return "unknown"
}

// --- internal helpers ---

func unstructuredItems(list *unstructured.UnstructuredList) []map[string]any {
	out := make([]map[string]any, len(list.Items))
	for i, item := range list.Items {
		out[i] = item.Object
	}
	return out
}
